AdblockPlusSafariExtension/AdblockPlus+Parsing.m

```diff
@@ -0,0 +1,308 @@
+                    withTemplate:@"\\\\$0"];
+  return h;
+}
+
++ (BOOL)mergeFilterListsFromURL:(NSURL *__nonnull)input
+        withWhitelistedWebsites:(NSArray<NSString *> *__nonnull)whitelistedWebsites
+                          toURL:(NSURL *__nonnull)output
+                          error:(NSError *__nullable *__nonnull)error
+{
+
+  NSInputStream *inputStream = [NSInputStream inputStreamWithURL:input];
+  NSOutputStream *outputStream = [NSOutputStream outputStreamWithURL:output append:NO];
+  yajl_gen g = NULL;
+  yajl_handle hand = NULL;
+  adblock_plus_context context = { 0, NULL };
+
+  @try {
+    [inputStream open];
+    [outputStream open];
+    g = yajl_gen_alloc(NULL);
+    yajl_gen_config(g, yajl_gen_beautify, 0);
+    yajl_gen_config(g, yajl_gen_validate_utf8, 0);
+    hand = yajl_alloc(&callbacks, NULL, (void *)&context);
+    yajl_config(hand, yajl_allow_comments, 0);
+    yajl_config(hand, yajl_dont_validate_strings, 1);
+    context.g = g;
+
+    // Write buffer to output stream
+    BOOL(^writeBuffer)(yajl_gen g) = ^BOOL(yajl_gen g) {
+      NSUInteger outputBufferLength;
+      const uint8_t *outputBuffer;
+      yajl_gen_get_buf(g, &outputBuffer, &outputBufferLength);
+      if (outputBufferLength > 0) {
+        NSInteger written = [outputStream write:outputBuffer maxLength:outputBufferLength];
+        if (written == -1) {
+          *error = outputStream.streamError;
+          return NO;
+        }
+      }
+      yajl_gen_clear(g);
+      return true;
+    };
+
+    // Get parser error
+    NSError *(^getParseError)(yajl_handle hand) = ^NSError *(yajl_handle hand) {
+      NSUInteger errorBufferLength = 256;
+      uint8_t errorBuffer[errorBufferLength];
+      unsigned char *error_string = yajl_get_error(hand, 1, errorBuffer, errorBufferLength);
+      NSString *errorString = [[NSString alloc] initWithCString:error_string encoding:NSASCIIStringEncoding];
+      return [NSError errorWithDomain:AdblockPlusErrorDomain code:0 userInfo:@{NSLocalizedDescriptionKey: errorString}];
+    };
+
+    // Read json file
+    const NSUInteger inputBufferLength = 256;
+    uint8_t inputBuffer[inputBufferLength];
+    NSInteger read;
+
+    while ((read = [inputStream read:inputBuffer maxLength:inputBufferLength]) != 0) {
+
+      yajl_status status = yajl_parse(hand, inputBuffer, read);
+      if (status != yajl_status_ok) {
+        *error = getParseError(hand);
+        return NO;
+      }
+
+      if (!writeBuffer(g)) {
+        return NO;
+      }
+    }
+
+    // Close parser
+    yajl_status status = yajl_complete_parse(hand);
+    if (status != yajl_status_ok) {
+      *error = getParseError(hand);
+      return NO;
+    }
+
+    // Write whitelisted websites
+    for (__strong NSString *website in whitelistedWebsites) {
+
+      NSDictionary *dictionary =
+      @{@"trigger": @{ @"url-filter": @".*", @"if-domain": @[website]},
+        @"action": @{ @"type": @"ignore-previous-rules" }
```

Comment:
At first try I moved added whitelisted page to url-filter, but QA reported that it does not work as expected. So that I used if-domain for whitelist which seems to work fine.